oauth/clientfactory/workspace_aware_client_factory.go

```diff
@@ -0,0 +1,122 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package clientfactory
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+
+	"github.com/codeready-toolchain/api/api/v1alpha1"
+	"k8s.io/client-go/rest"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+const wsApiPath = "/apis/toolchain.dev.openshift.com/v1alpha1/workspaces"
+
+var (
+	errUnableToParseWorkspaceResponse = errors.New("unable to parse response from workspace API requested for namespace")
+	errWorkspaceNotFound              = errors.New("target workspace not found for namespace")
+)
+
+// WorkspaceAwareK8sClientFactory is a K8S client factory, which is authenticates on server via user token,
+// but it uses the custom API server URL and sets the workspace path in the requests to the cluster,
+// by consuming the namespace in the context when new client instance is created.
+type WorkspaceAwareK8sClientFactory struct {
+	ClientOptions *client.Options
+	RestConfig    *rest.Config
+	ApiServer     string
+	HTTPClient    rest.HTTPClient
+}
+
+func (w WorkspaceAwareK8sClientFactory) CreateClient(ctx context.Context) (client.Client, error) {
+	namespace, ok := NamespaceFromContext(ctx)
+	if !ok { // no namespace, return simple client
+		return doCreateClient(w.RestConfig, *w.ClientOptions)
+	}
+	lg := log.FromContext(ctx)
+
+	wsName, err := fetchWorkspaceName(ctx, w.ApiServer, namespace, w.HTTPClient)
+	if err != nil {
+		lg.Error(err, "failed to fetch workspace name via API:", "api", w.RestConfig.Host, "workspace", wsName, "error", err.Error())
+		return nil, fmt.Errorf("unable to fetch workspace name via API: %w", err)
+	}
+	w.RestConfig.Host, err = url.JoinPath(w.RestConfig.Host, "workspaces", wsName)
+	if err != nil {
+		lg.Error(err, "failed to create the K8S API path", "api", w.RestConfig.Host, "workspace", wsName)
+		return nil, fmt.Errorf("failed to create the K8S API path from URL %s and workspace %s: %w", w.RestConfig.Host, wsName, err)
+	}
+	return doCreateClient(w.RestConfig, *w.ClientOptions)
+}
+
+// fetchWorkspaceName finds the workspace to which the given namespace is belongs
+func fetchWorkspaceName(ctx context.Context, apiServer, namespace string, httpClient rest.HTTPClient) (string, error) {
+	wsList, err := fetchAllAvailableWorkspaces(ctx, apiServer, namespace, httpClient)
+	if err != nil {
+		return "", fmt.Errorf("unable to list available workspaces foe namespace %s, error:%w", namespace, err)
+	}
+	for _, ws := range wsList.Items {
+		for _, ns := range ws.Status.Namespaces {
+			if ns.Name == namespace {
```

Comment:
so how workspaces works? Can for example every workspace have `default` namespace? How will this behave then? Does this `WorkspaceAwareK8sClientFactory` expect some exact use of workspaces? Is it tested?

Reply:
i would be glad if someone answer those questions. so far nobody knows.

Reply:
yes, but what if 2 workspaces have namespace with same name? how do we know which one to use ?

Reply:
ok, can we document it somehow? It is a lot of internal information from that we're making assumptions in the code.